DOCSP-32067: Update Create page for Kotlin (#3039)
mongodb/docs-realm: source/sdk/kotlin/realm-database/crud/create.txt

   :depth: 2
   :class: singlecol

This page describes how to create a new object in a local or synced realm 
using the Kotlin SDK. To learn more about Realm objects and how to define them, 
refer to :ref:`<kotlin-object-type>`.

You can create objects whose object type is included in the 
realm schema when you open the realm.
For more information, refer to 

> :ref:`<kotlin-open-a-realm>` or :ref:`<kotlin-open-a-synced-realm>`.

Author: Yeah, this one's especially jarring. Updated

.. note:: Write to a Synced Realm

   The syntax to write a new object to a realm is the same for a local or
   a synced realm. However, there are additional considerations that determine
   whether the write operation in a synced realm is successful. For more 
   information, refer to :ref:`<kotlin-write-synced-realm>`.

.. _kotlin-write-transactions:
.. _kotlin-open-a-transaction:

> Write Transactions

Reviewer: Are you planning on putting this section at the top of the Update and Delete pages too? Should it be an include?

Author: I'm working on those pages now. On the fence on how much to duplicate vs. link out. Will play around with it after this is merged in

within a transaction.

A **transaction** is a list of read and write operations that
Realm treats as a single indivisible operation: either *all* of the
operations succeed or *none* of the operations in the transaction 
take effect.

Realm represents each transaction as a callback function
that contains zero or more read and write operations. To run
a transaction, you define a transaction callback and pass it to
the realm's 
`write() <{+kotlin-local-prefix+}io.realm.kotlin/-realm/write.html>`__ 
method. Within this callback, you are
free to create, read, update, and delete on the realm. 

A realm allows only one open write transaction at a time. Realm
blocks other writes on other threads until the open
transaction is complete. Consequently, there is no race
condition when reading values from the realm within a
transaction.

When you are done with your transaction, Realm either
commits it or cancels it:

- When Realm commits a transaction, Realm writes
  all changes to disk. For synced realms, the SDK queues the change
  for synchronization with the backend.
- When Realm cancels a write transaction or an operation in
  the transaction causes an error, all changes are discarded

>   (or "rolled back").

Author: Nope. This was existing language that I didn't look too hard at. Removed.

.. note:: Frozen Objects
   
   Objects returned from a write closure become frozen objects when the 
   write transaction completes. For more information, refer to 
   :ref:`<kotlin-frozen-architecture>`.

.. _kotlin-managed-vs-unmanaged-objects:

Managed vs. Unmanaged Objects
-----------------------------

Realm APIs may refer to objects as managed or unmanaged. When you create 
a Realm object with the Kotlin SDK, it is unmanaged by the realm. 
You can create managed instance by copying an unmanaged object to a realm.
To create an unmanaged copy of a managed object, refer to the

> :ref:`<kotlin-create-unmanaged-copy>` section. 

Reviewer: Do we want to specify that section is on this page?
```suggestion
:ref:`<kotlin-create-unmanaged-copy>` section on this page. 
```

> :ref:`<kotlin-create-unmanaged-copy>` section. 

Author: Updated

- **Managed objects** are Realm objects that persist in a realm. Managed objects 
  can only be accessed from an open realm. They can be automatically updated 
  with changes within write transactions as long as that realm remains open. 
  Managed objects are tied to the realm instance from which they originated 
  and cannot be written to another realm.

  You can use Realm APIs with managed objects, create relationships with 
  other objects, and observe managed Realm objects for changes. 

- **Unmanaged objects** are instances of Realm objects that
  behave like normal Kotlin objects, but they are not persisted in a realm.
  All Realm objects are unmanaged until you write them to a realm.
  You cannot use Realm APIs with unmanaged objects or observe them for changes.

.. tip:: 

   You can check if an object is managed with the 
   `isManaged() <{+kotlin-local-prefix+}io.realm.kotlin.ext/is-managed.html>`__  
   method.

> Create a New Managed Object

Author: I like that. Updated

---------------------------

After you :ref:`<kotlin-define-a-new-object-type>` in your object model 
and include it in your realm schema when you open the realm,
you can reference and write those objects in your realm.

.. important:: Object Types Must Be in Your Schema

   You can only write objects whose object type 
   is included in the realm schema. If you try to 
   reference or write an object of an object type 
   that isn't in your schema, Realm will return a 
   schema validation error.

> To persist a new managed object to a realm:

Reviewer: You can't really persist a managed object, can you? Isn't a managed object something that's already been added to the realm (and been persisted)?

To make this clearer, I think we could rename this section (see my question above) or change this sentence to match the current title. Something like "To create a new managed object".

Author: Derp. Good catch

> 1. Instantiate a new object instance with the class constructor. You

Reviewer: It seems that steps 1 and 2 should be swapped. Or maybe it doesn't matter if the object is instantiated outside the write transaction?

The Create a Realm Object example shows it inside the write transaction.

#. Open a write transaction with `realm.write()
   <{+kotlin-local-prefix+}io.realm.kotlin/-realm/write.html>`__ or
   `realm.writeBlocking()
   <{+kotlin-local-prefix+}io.realm.kotlin/-realm/write-blocking.html>`__
   to instantiate the unmanaged object.

#. Pass the new object instance to `copyToRealm()
   <{+kotlin-local-prefix+}io.realm.kotlin/-mutable-realm/copy-to-realm.html>`__
   to persist the object data to the realm. This method returns a
   live managed instance of the object. 

   .. important:: Asymmetric Objects Use Insert()
   
      Asymmetric objects are special write-only objects that do not 
      persist to the realm. They *do not* use ``copyToRealm()``. 
      Instead, you pass the asymmetric object instance to the 
      ``insert()`` extension method within a write transaction.

>       Refer to the :ref:`<kotlin-create-asymmetric-object>` section 
>       for more information.

Reviewer: Do we want to specify that section is on this page?
```suggestion
      Refer to the :ref:`<kotlin-create-asymmetric-object>` section 
      on this page for more information.
```

For more information about the ``MutableRealmInt`` type, refer to 
:ref:`<kotlin-mutablerealmint>`.

In the following example, we instantiate a new ``Frog`` object with a 
``fliesEaten`` property and pass an initial value to 
``MutableRealmInt.create()``:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-mutablerealm-property.kt
   :language: kotlin

.. To learn how to increment, decrement, and set the counter value after 
.. you create the object, refer to :ref:`<kotlin-update-mutablerealm>`.

Create a RealmAny (Mixed) Property
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create a new object instance with a polymorphic 
`RealmAny <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-any/index.html>`__  
property, instantiate an object and pass an initial value of a 
supported type to the ``RealmAny`` property using 
`RealmAny.create() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-any/-companion/create.html>`__. 
For a list of the value types that ``RealmAny`` can hold, refer to 
:ref:`<kotlin-realmany>`.

In the following example, we instantiate a new ``Frog`` object with a 

> ``favoriteFood`` list of ``RealmAny`` type and pass the initial values to 

Reviewer: The code snippet's property is named ``favoriteThings``, not ``favoriteFood``.

Author: thanks for catching

``RealmAny.create()``:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-realmany-property.kt
   :language: kotlin

> After you create the object, you can *must* know the stored value type 

Reviewer: ```suggestion
After you create the object, you *must* know the stored value type 
```

to work with the ``RealmAny`` property. 

.. To learn how to work with ``RealmAny`` properties after 
.. you create the object, refer to :ref:`<kotlin-update-realmany>`.

.. _kotlin-create-a-collection:

Create Collection Properties
----------------------------

Depending on how you define your object type, you might have properties 
that are defined as one of the following 
supported :ref:`<kotlin-collection-types>`:

> ``RealmList``, ``RealmSet``, and ``RealmDictionary``. For more information, 

Reviewer: Consider breaking into an unordered list for scanability.

Author: Updated

refer to :ref:`<kotlin-define-collections>`.

Collections are mutable: you can add and remove elements in a collection
within a write transaction. 

> Collections can contain both managed and unmanaged objects. When a 

Reviewer: ```suggestion
Collections can contain both managed and unmanaged objects. When
```

supported type to the ``RealmList`` property. For a list of the value types
that ``RealmList`` can hold, refer to :ref:`<kotlin-realm-list>`. 

You can instantiate an unmanaged list with `realmListOf() <{+kotlin-local-prefix+}io.realm.kotlin.ext/realm-list-of.html>`__ 
or pass elements to the list using
`list.add() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-list/index.html#882550416%2FFunctions%2F878332154>`__, 
`list.addAll() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-list/index.html#-4875109%2FFunctions%2F878332154>`__,
or `list.set() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-list/set.html>`__.
The list is unmanaged until you copy it to the realm.

In the following example, we instantiate a new ``Frog`` object with 
initial values for several ``RealmList`` properties:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-realm-list.kt
   :language: kotlin

Create a RealmSet Property
~~~~~~~~~~~~~~~~~~~~~~~~~~

To create a new object instance with a 
`RealmSet 
<{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-set/index.html>`__  
property, instantiate an object and pass any values of a 

> supported type to the ``RealmSet`` property. For a list of valid value types

Reviewer: > valid value types

Should this be either "valid values" or "valid types"?

that ``RealmSet`` can hold, refer to :ref:`<kotlin-realm-set>`. 

You can instantiate an unmanaged set with `realmSetOf() <{+kotlin-local-prefix+}io.realm.kotlin.ext/realm-list-of.html>`__ 
or pass elements to the set using 
`set.add() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-set/index.html#-153241610%2FFunctions%2F878332154>`__
or  
`set.addAll() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-set/index.html#-800009087%2FFunctions%2F878332154>`__. 
The set is unmanaged until you copy it to the realm.

In the following example, we instantiate a new ``Frog`` object with
initial values for ``favoriteSnacks`` and ``favoriteWeather`` set 
properties:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-realm-set.kt
   :language: kotlin

Create a Dictionary Property
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create a new object instance with a 
`RealmDictionary <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-dictionary/index.html>`__ 
property, instantiate an object and pass any key-value pairs of a 
supported type to the ``RealmDictionary`` property. 
``RealmDictionary`` only accepts a ``String`` key, but values may be 

> non-string types. For a list of valid value types, refer to 

Reviewer: Same as previous comment:

> valid value types

Should this be either "valid values" or "valid types"?

Author: updated

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.percent-encode-disallowed-characters.kt
   :language: kotlin

> Create a Relationship

Reviewer: Does it make sense to use a similar title as the other section titles: `Create a Relationship Property`?

Author: I ended up removing `Property` in these Relationship sections to decrease the heading lengths and improve the on-page TOC readability. I can add it back if it's confusing

Reviewer: Nah, I don't think it's a big more. More the writer in me that wants parallelism. I don't think it will really matter to readers. 😄

---------------------

Depending on how you define your object type, you might have properties 
that reference another Realm object. This can be a to-one, to-many, or
inverse :ref:`relationship <kotlin-relationships>`. 
For more information on defining relationships in your object model, 
refer to :ref:`<kotlin-define-a-relationship>`.

You can also embed one Realm object directly within another to create 
a nested data structure with an ``EmbeddedRealmObject`` type. 
To create a relationship with an embedded object, refer to the 

> :ref:`<kotlin-create-an-embedded-object>` section.

Reviewer: Do we want to specify that section is on this page?
```suggestion
:ref:`<kotlin-create-an-embedded-object>` section on this page.
```

Create a To-One Relationship
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create a new object instance with a 
:ref:`to-one relationship <kotlin-to-one-relationship>`

> property, instantiate both objects and pass the referenced object in the

Reviewer: ```suggestion
property, instantiate both objects and pass the referenced object to the
```

relationship property. 

In the following example, we instantiate a new ``Frog`` object with a 
``favoritePond`` property that references a ``Pond`` object and a 
``bestFriend`` property that references another ``Frog`` object:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-to-one-realm-relationship.kt
   :language: kotlin

Create a To-Many Relationship
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create a new object instance with a 
:ref:`to-many relationship <kotlin-to-many-relationship>`

> property, instantiate all objects and pass any referenced objects in the

Reviewer: ```suggestion
property, instantiate all objects and pass any referenced objects to the
```

relationship collection property. 

In the following example, we instantiate a new ``Forest`` object with a 
``frogsThatLiveHere`` property that references a set of ``Frog`` 

> objects and a ``nearBy`` property that references a list of 

Reviewer: Code snippet property is ``nearbyPonds`` instead of ``nearBy``.

Author: thanks for catching!

``Pond`` objects:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-to-many-realm-relationship.kt
   :language: kotlin

Create an Inverse Relationship
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To create a new object instance with an 
:ref:`inverse relationship <kotlin-inverse-relationships>`
property, instantiate the parent object and pass any referenced child

> objects in the backlink collection property. 

Reviewer: ```suggestion
objects to the backlink collection property. 
```

In the following example, we instantiate a new ``User`` object with a 
backlinks ``posts`` property that references a list of ``Post`` objects:

.. literalinclude:: /examples/generated/kotlin/CreateTest.snippet.create-inverse-realm-relationship.kt
   :language: kotlin

After you create the object, you can access the backlinks collection
property to get the child objects. You *cannot* directly modify 
the backlink itself. 

.. For more information, refer to :ref:`<kotlin-update-inverse-relationships>`.

.. _kotlin-create-unmanaged-copy:

> Create an Unmanaged Copy of a RealmObject or Collection

Reviewer: ```suggestion
Create an Unmanaged Copy of a Realm Object or Collection
```
